src/components/networked-hand-controls.js

```diff
@@ -153,38 +113,284 @@ AFRAME.registerComponent('networked-hand-controls', {
+        // this.el.setAttribute(this.data.controllerComponent, 'model', true)
+
+        // so we first remove the controller component as best we can
+        if (this.el.components[this.data.controllerComponent]) {
+          // there is generally no remove() specified, but they do have pause(), which is close enough:
+          try {
+            this.el.components[this.data.controllerComponent].pause();
+            this.el.components[this.data.controllerComponent].remove();
+          } catch(e) {
+            console.error("NAF: possible problem while updating handModelStyle to controller", e)
+          }
+          this.el.removeAttribute(this.data.controllerComponent);
+        }
+
+        if (!this.local) {
+          this.injectRemoteControllerModel();
+        }
+        else {
+          this.addControllerComponents(true);
+          // controllerconnected event won't fire again, so we have to call this manually:
+          this.el.components[this.data.controllerComponent].injectTrackedControls({profiles:JSON.parse(this.data.webxrControllerProfiles)}) // have to call this manually, since the controller isn't newly connected
+        }
+      }
+    }
+
+    if (oldData.color &&
+        this.data.handModelStyle !== this.str.controller &&
+        oldData.color !== this.data.color) {
+      this.updateHandMeshColor();
+    }
     
-    if (!this.local && (this.data.gesture != oldData.gesture)) {
-      this.handleGesture(this.data.gesture, oldData.gesture);
+    // this block is for receiving+handling networked hand gestures and button events for controller model updates
+    // and for initial injection of remote controller model
+    if (!this.local) {
+      if (this.data.handModelStyle !== this.str.controller && this.data.gesture !== oldData.gesture) {
+        this.handleGesture(this.data.gesture, oldData.gesture);
+      }
+      else if (this.data.handModelStyle === this.str.controller) {
+        if (!this.injectedController && this.data.controllerComponent && this.data.webxrControllerProfiles[0]) {
+          this.injectRemoteControllerModel();
+        }
+        if (oldData.controllerEvent && oldData.controllerEvent !== this.data.controllerEvent) {
+          const controllerEvent = JSON.parse(this.data.controllerEvent);
+          this.el.components[this.data.controllerComponent][controllerEvent.type === 'thumbstickmoved' ? 'onThumbstickMoved' : 'onButtonChanged'](controllerEvent);
+        }
+        if (oldData.controllerModelUpdate && (
+          oldData.controllerModelUpdate[0] !== this.data.controllerModelUpdate[0] ||
+          oldData.controllerModelUpdate[1] !== this.data.controllerModelUpdate[1] 
+        )) {
+          this.el.components[this.data.controllerComponent].updateModel(...this.data.controllerModelUpdate)
+        }
+      }
     }
   },
 
   remove() {
-    this.el.removeObject3D('mesh');
+    this.el.removeObject3D(this.str.mesh);
+    this.removeEventListeners();
```

Comment:
Are you sure you want `this.removeEventListeners();` here? It seems wrong. `this.removeEventListeners();` is already in `pause()` for local player symmetrically with the `this.addEventListeners()` for the local player in `play()`

Reply:
yeah, that's correct. I hadn't touched A-Frame almost at all since last fall until the past month or so. Not sure if I forgot or never knew, but I've either learned or re-learned from you about the removeAttribute calling remove which calls pause pattern. when I started I assumed it was just incomplete. makes sense, though.

thanks for the review, I think everything I wrote would have worked fine either way, but I definitiely gained knowledge about writing better a-frame code.

Reply:
Thank you for your patience. I'm glad you learned some things from this review.

I also relearned yesterday that when adding a component, it runs `init` and then `update`!, then `play`.
See https://github.com/aframevr/aframe/blob/795edfd65841d54aa2b186a02f660ffee606b50c/src/core/component.js#L306-L336
I had a code in `update` that was replacing an outfit material for an avatar, and I saw it did the logic twice.
My code was this:
```
schema: {
    outfit: { default: "" },
},
update: function (oldData) {
    if (!oldData) return; // Normally always a {} if it's an object based component when adding the component locally, but can be undefined some times when NAF is creating it.
 
    if (oldData.outfit !== data.outfit) {
      this.loadOutfit();
    }
}
```
To fix the issue I just added an extra condition:
```
update: function (oldData) {
    if (!oldData) return; // Normally always a {} if it's an object based component when adding the component locally, but can be undefined some times when NAF is creating it.
 
    if (oldData.outfit && oldData.outfit !== data.outfit) {
      this.loadOutfit();
    }
}
```
That works because I know I'm setting the outfit when creating the component, I don't want a naked avatar. :)